pkg/cluster/components/node.go

```diff
@@ -0,0 +1,103 @@
+package components
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/kubernetes-incubator/bootkube/pkg/node"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+// NodeUpdater is responsible for updating nodes via
+// annotations which are then handled by the node-agent.
+type NodeUpdater struct {
+	// client is an API Server client.
+	client clientset.Interface
+	// nodeStore is cache of node backed by an informer.
+	nodes cache.StoreToNodeLister
+	// node is the node this updater is responsible for.
+	node *api.Node
+}
+
+func NewNodeUpdater(client clientset.Interface, node *api.Node, nodes cache.StoreToNodeLister) (*NodeUpdater, error) {
+	return &NodeUpdater{
+		client: client,
+		nodes:  nodes,
+		node:   node,
+	}, nil
+}
+
+func (nu *NodeUpdater) Name() string {
+	return nu.node.Name
+}
+
+// Priority for Nodes should be such that they
+// get updated last, and rolled back first.
+func (nu *NodeUpdater) Priority() int {
+	return 100
+}
+
+// Version returns the highest version of any
+// Node in the store.
+func (nu *NodeUpdater) Version() (*Version, error) {
+	ni, exists, err := nu.nodes.Get(nu.node)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("could not find Node %s in store", nu.Name())
+	}
+	n, ok := ni.(*api.Node)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type returned from Node store, expected *v1.Node, got %T", ni)
+	}
+	versionString, ok := n.Annotations[node.CurrentVersionAnnotation]
+	if !ok {
+		return nil, fmt.Errorf("no version annotation for Node %s", n.Name)
+	}
+	return ParseVersionFromImage(versionString)
+}
+
+// UpdateToVersion will update the Node to the given version.
+func (nu *NodeUpdater) UpdateToVersion(v *Version) (bool, error) {
+	n, err := nu.client.Core().Nodes().Get(nu.Name())
+	if err != nil {
+		return false, err
+	}
+	// First step: update the annotation on the Node object. This will
+	// trigger the node-agent running on that node to update the Node.
+	if n.Annotations == nil {
+		n.Annotations = make(map[string]string)
+	}
+	if n.Annotations[node.CurrentVersionAnnotation] == v.image.String() {
+		return false, nil
+	}
+	n.Annotations[node.DesiredVersionAnnotation] = v.image.String()
+	_, err = nu.client.Core().Nodes().Update(n)
+	if err != nil {
+		return false, err
+	}
+	// Second step: wait until the node-agent has updated the Node.
+	err = wait.Poll(time.Second, 10*time.Minute, func() (bool, error) {
```

Comment:
nit: do we really need this long of timeout? Also can we change these to be constants (e.g. nodeUpdateTimeout, deploymentUpdateTimeout, etc.)